[FLINK-16124] [kinesis] Support Java Kinesis Ingress / Egress

tzulitai opened the pull request:
This PR adds support for a first-class Java ingress and egress for AWS Kinesis streams.

This PR provides the slimmest possible setup for the initial support.
For example, only the most commonly used ways (w.r.t. AWS frequent users) for configuring AWS credentials / AWS region are supported.
Some more sophisticated features (e.g. authenticating via AWS STS role assumption) can be added in the future as follow-ups.

---

### User API

Please see the following for example usage:
- `KinesisIngressBuilderTest` added in 812a528
- `KinesisEgressBuilderTest` added in 3086f33

As you'll see, the out-of-box usage is very simple.
For ingresses, only the consumed stream name + deserializer is required.
For egresses, only the serializer is required.

AWS connection configuration, by default  is all picked up via AWS's default provider chains (which is a large majority of how AWS users connect to AWS services).
Please see the Javadocs of `KinesisIngressBuilder.withAwsRegion` and `KinesisIngressBuilder.withAwsCredentials` for further details.

---

### Changelog

The major additions are as follows:

- 812a528 SDK classes for ingress
- b8cd5f0 Runtime classes / Flink source provider for ingress
- 65434ac Binds the source provider for ingress
- 3086f33 SDK classes for egress
- 7a62ee0 Runtime classes / Flink sink provider for egress
- 94b596c Binds the sink provider for egress

---

### Verification

For now, I've already manually tested this against an actual AWS Kinesis stream.
I'm planning to add a follow-up E2E test that tests against [Kinesalite](https://github.com/mhart/kinesalite), a widely used mock implementation of the AWS Kinesis service interface for testing purposes.

igalshilman commented on statefun-flink/statefun-flink-io-bundle/src/main/java/org/apache/flink/statefun/flink/io/kinesis/AwsAuthConfig.java:
```diff
@@ -0,0 +1,128 @@
+      Properties properties, AwsRegion.CustomEndpointAwsRegion customEndpoint) {
+    URL url;
+    try {
+      url = new URL(customEndpoint.serviceEndpoint());
```
You may want to switch to a URI
If you are only interested in the host.
URL requires a DNS lookup.

igalshilman commented on statefun-flink/statefun-flink-io-bundle/src/main/java/org/apache/flink/statefun/flink/io/kinesis/AwsAuthConfig.java:
```diff
@@ -0,0 +1,128 @@
+    }
+  }
+
+  private static void setAwsCredentialProperties(
```
This could be a nice method to test.

tzulitai commented:
Thanks a lot for the review @igalshilman!
I've added unit tests as you suggested, and also reworked the serialization interfaces as we discussed offline.

Since the follow-up changes are mostly trivial, I will now proceed to merge this PR ...
